[clang-tidy] add new check readability-enum-initial-value (#86129)
llvm/llvm-project: clang-tools-extra/docs/clang-tidy/checks/readability/enum-initial-value.rst

.. title:: clang-tidy - readability-enum-initial-value

readability-enum-initial-value
==============================

Detects explicit initialization of a part of enumerators in an enumeration, and
relying on compiler to initialize the others.

> It causes readability issues and reduces the maintainability. When adding new

Reviewer: no need for " readability issues and reduces the maintainability", just say that this is about consistency.

.. title:: clang-tidy - readability-enum-initial-value

readability-enum-initial-value
==============================

Detects explicit initialization of a part of enumerators in an enumeration, and

> relying on compiler to initialize the others.

Reviewer: Maybe:
```
Enforces consistent style for enumerators' initialization, covering three styles: none, first only, or all initialized explicitly.
```

When adding new enumerations, inconsistent initial value will cause potential
enumeration value conflicts.

In an enumeration, the following three cases are accepted. 
1. none of enumerators are explicit initialized.
2. the first enumerator is explicit initialized.
3. all of enumerators are explicit initialized.

.. code-block:: c++

  // valid, none of enumerators are initialized.
  enum A {
    e0,
    e1,
    e2,
  };

  // valid, the first enumerator is initialized.
  enum A {
    e0 = 0,
    e1,
    e2,
  };

  // valid, all of enumerators are initialized.
  enum A {
    e0 = 0,
    e1 = 1,
    e2 = 2,
  };

  // invalid, e1 is not explicit initialized.
  enum A {
    e0 = 0,
    e1,
    e2 = 2,
  };

Options
-------

.. option:: AllowExplicitZeroFirstInitialValue

>   If set to `false`, explicit initialized first enumerator is not allowed.

Reviewer: ```suggestion
  If set to `false`, the first enumerator must not be explicitly initialized.
```

  See examples below. Default is `true`.

  .. code-block:: c++

    enum A {
      e0 = 0, // not allowed if AllowExplicitZeroFirstInitialValue is false
      e1,
      e2,
    };

> .. option:: AllowExplicitLinearInitialValues

Reviewer: Consider using `sequential` instead of `linear`.
